lib/url_to_subscriber_list_criteria.rb

```diff
@@ -0,0 +1,149 @@
+require 'uri'
+
+class UrlToSubscriberListCriteria
+  MISSING_LOOKUP = "*** MISSING KEY ***"
+  EMAIL_SUPERTYPE = "email_document_supertype"
+  GOVERNMENT_SUPERTYPE = "government_document_supertype"
+  class UnprocessableUrl < StandardError; end
+
+  attr_reader :missing_lookup
+
+  def initialize(url, static_data = StaticData)
+    @url = URI.parse(url.strip)
+    @static_data = static_data
+    @missing_lookup = false
+  end
+
+  def convert
+    hash = map_url_to_hash.dup
+    if hash["links"]
+      links = hash["links"].each_with_object({}) do |(key, values), result|
+        result[key] = values.map { |value| lookup_content_id(key, value) }
+      end
+      hash.merge!("links" => links)
+    end
+    hash
+  end
+
+  def map_url_to_hash
+    @map_url_to_hash ||= begin
+      result = if @url.path == "/government/feed" || @url.path == "/government/feed.atom"
+                 { 'links' => from_params }
+               elsif (result = @url.path.match(%r{^/government/world/(.*)\.atom$}))
+                 { "links" => from_params.merge("world_locations" => [result[1]]) }
+               elsif (result = @url.path.match(%r{^/government/topical-events/(.*)\.atom$}))
+                 { "links" => from_params.merge("topical_events" => [result[1]]) }
+               elsif (result = @url.path.match(%r{^/government/people/(.*)\.atom$}))
+                 { "links" => from_params.merge("people" => [result[1]]) } # this is not yet implemented
+               elsif (result = @url.path.match(%r{^/government/organisations/(.*)\.atom$}))
+                 { "links" => from_params.merge("organisations" => [result[1]]) }
+               elsif (result = @url.path.match(%r{^/government/topics/(.*)\.atom$}))
+                 { "links" => from_params.merge(topic_map([result[1]]) => [result[1]]) }
+               elsif (result = @url.path.match(%r{^/government/ministers/(.*)\.atom$}))
+                 { "links" => from_params.merge("roles" => [result[1]]) } # this is not yet implemented
+
+
+               elsif @url.path.match(%r{^/government/statistics\.atom$})
+                 { "links" => from_params, GOVERNMENT_SUPERTYPE => "statistics" }
+               elsif @url.path.match(%r{^/government/publications\.atom$})
+                 { "links" => from_params, EMAIL_SUPERTYPE => "publications" }
+               elsif @url.path.match(%r{^/government/announcements\.atom$})
+                 { "links" => from_params, EMAIL_SUPERTYPE => "announcements" }
+
+               else
+                 raise UnprocessableUrl, @url.to_s
+               end
+
+      if result.fetch("links", {})["publication_filter_option"]
+        result[GOVERNMENT_SUPERTYPE] = result["links"].delete("publication_filter_option")
+      end
+      if result.fetch("links", {})["announcement_filter_option"]
+        result[GOVERNMENT_SUPERTYPE] = result["links"].delete("announcement_filter_option")
```

Comment:
OK, let's not worry about that - I don't think there are any bad ones in the dataset at the moment.